File: javascript/examples/cost_query_example.js
/**
 * Cost Query Example
 * 使用kubeconfig进行身份验证，查询Kubernetes集群中的成本数据
 */

const { ApiClient, DefaultApi } = require('../dist/index.js');
const fs = require('fs');
const path = require('path');

/**
 * 查找kubeconfig文件
 * @returns {string|null} kubeconfig文件路径
 */
function findKubeconfigFile() {
    // 尝试KUBECONFIG环境变量
    const kubeconfigEnv = process.env.KUBECONFIG;
    if (kubeconfigEnv && fs.existsSync(kubeconfigEnv)) {
        return kubeconfigEnv;
    }

    // 尝试默认路径
    const homeDir = process.env.HOME || process.env.USERPROFILE;
    const defaultPath = path.join(homeDir, '.kube', 'config');
    if (fs.existsSync(defaultPath)) {
        return defaultPath;
    }
    
    return null;
}

/**
 * 查询成本数据
 */
async function queryCostData() {
    console.log('============================================================');
    console.log('JavaScript SDK Kubeconfig Cost Query Example');
    console.log('============================================================');

    const kubeconfigPath = findKubeconfigFile();
    if (!kubeconfigPath) {
        console.error('No kubeconfig file found. Please ensure KUBECONFIG environment variable is set or file exists at ~/.kube/config');
        process.exit(1);
    }
    console.log(`Found kubeconfig file: ${kubeconfigPath}`);

    console.log('Starting cost data query...');
    try {
        // 1. 使用kubeconfig和代理创建API客户端
        // 代理路径与kubectl命令中的路径一致
        const proxyPath = '/api/v1/namespaces/kube-system/services/ack-metrics-adapter-api-service:8080/proxy';
        const apiClient = ApiClient.newApiClientWithKubeconfigAndProxy(
            kubeconfigPath,
            proxyPath
        );
        console.log('Client created successfully!');
        console.log(`Server URL: ${apiClient.basePath}`);

        // 2. 创建API实例
        const defaultApi = new DefaultApi(apiClient);
        console.log('API instance created successfully');

        // 3. 定义查询参数
        const window = '1h';
        const filter = 'namespace:"kube-system"+controllerKind:"ReplicaSet"+label[app]:"ack-cost-exporter"';
        
        console.log('Querying cost data for ack-cost-exporter in kube-system namespace...');
        console.log(`Request URL: ${apiClient.basePath}/v2/cost?window=${window}&filter=${encodeURIComponent(filter)}`);

        // 4. 发送API请求
        const opts = {
            filter: filter
        };
        
        const response = await new Promise((resolve, reject) => {
            defaultApi.getCost(window, opts, (error, data, response) => {
                if (error) {
                    reject(error);
                } else {
                    resolve(data);
                }
            });
        });

        console.log('Successfully retrieved cost data');
        console.log('Response data structure:');

        if (response && response.data) {
            console.log(`   - Number of data entries: ${response.data.length}`);
            
            // Display each time range's data
            for (let i = 0; i < response.data.length; i++) {
                const dataItem = response.data[i];
                console.log(`\n   Time Range ${i + 1}:`);
                
                if (dataItem && typeof dataItem === 'object') {
                    // Display each pod's cost data in this time range
                    for (const [resourceName, allocation] of Object.entries(dataItem)) {
                        if (allocation && typeof allocation === 'object') {
                            console.log(`\n     Pod: ${resourceName}`);
                            
                            // Display basic allocation info
                            if (allocation.name) {
                                console.log(`       Name: ${allocation.name}`);
                            }
                            if (allocation.start) {
                                console.log(`       Start: ${allocation.start}`);
                            }
                            if (allocation.end) {
                                console.log(`       End: ${allocation.end}`);
                            }
                            
                            // Display resource usage
                            if (allocation.cpuCoreRequestAverage !== undefined) {
                                console.log(`       CPU Request Average: ${allocation.cpuCoreRequestAverage.toFixed(2)} cores`);
                            }
                            if (allocation.cpuCoreUsageAverage !== undefined) {
                                console.log(`       CPU Usage Average: ${allocation.cpuCoreUsageAverage.toFixed(2)} cores`);
                            }
                            if (allocation.ramByteRequestAverage !== undefined) {
                                console.log(`       RAM Request Average: ${(allocation.ramByteRequestAverage / (1024 * 1024)).toFixed(2)} MB`);
                            }
                            if (allocation.ramByteUsageAverage !== undefined) {
                                console.log(`       RAM Usage Average: ${(allocation.ramByteUsageAverage / (1024 * 1024)).toFixed(2)} MB`);
                            }
                            
                            // Display cost information
                            if (allocation.cost !== undefined) {
                                console.log(`       Cost: ${allocation.cost.toFixed(4)}`);
                            }
                            if (allocation.costRatio !== undefined) {
                                console.log(`       Cost Ratio: ${(allocation.costRatio * 100).toFixed(2)}%`);
                            }
                            if (allocation.customCost !== undefined) {
                                console.log(`       Custom Cost: ${allocation.customCost.toFixed(4)}`);
                            }
                            
                            // Display properties if available
                            if (allocation.properties) {
                                console.log(`       Properties:`);
                                if (allocation.properties.namespace) {
                                    console.log(`         Namespace: ${allocation.properties.namespace}`);
                                }
                                if (allocation.properties.node) {
                                    console.log(`         Node: ${allocation.properties.node}`);
                                }
                                if (allocation.properties.controller) {
                                    console.log(`         Controller: ${allocation.properties.controller}`);
                                }
                                if (allocation.properties.controllerKind) {
                                    console.log(`         Controller Kind: ${allocation.properties.controllerKind}`);
                                }
                                if (allocation.properties.cluster) {
                                    console.log(`         Cluster: ${allocation.properties.cluster}`);
                                }
                            }
                        }
                    }
                }
            }
        } else {
            console.log('   - No matching resources found');
        }

    } catch (error) {
        console.error('Query failed:', error.message);
        if (error.response) {
            console.error('HTTP status code:', error.response.status);
            console.error('Response body:', error.response.body);
        }
        process.exit(1);
    } finally {
        console.log('============================================================');
        console.log('Example completed successfully!');
    }
}

// 运行示例
if (require.main === module) {
    queryCostData().catch(error => {
        console.error('Example execution failed:', error);
        process.exit(1);
    });
}

module.exports = { queryCostData, findKubeconfigFile };

File: java/examples/CostQueryExample.java
package examples;

import org.openapitools.client.ApiClient;
import org.openapitools.client.ApiException;
import org.openapitools.client.Configuration;
import org.openapitools.client.api.DefaultApi;
import org.openapitools.client.model.AllocationSetRange;

import java.io.IOException;

/**
 * Cost Query Example
 */
public class CostQueryExample {
    
    public static void main(String[] args) {
        System.out.println("============================================================");
        System.out.println("Java SDK Kubeconfig Cost Query Example");
        System.out.println("============================================================");
        
        try {
            // Create API client with kubeconfig and proxy
            String proxyPath = "/api/v1/namespaces/kube-system/services/ack-metrics-adapter-api-service:8080/proxy";
            ApiClient apiClient = Configuration.newApiClientWithKubeconfigAndProxy(null, proxyPath);
            
            System.out.println("Client created successfully!");
            System.out.println("Server URL: " + apiClient.getBasePath());
            
            // Create API instance
            DefaultApi defaultApi = new DefaultApi(apiClient);
            System.out.println("API instance created successfully");
            
            // Query cost data
            System.out.println("Querying cost data for ack-cost-exporter in kube-system namespace...");
            
            // Make the API call
            AllocationSetRange response = defaultApi.getCost(
                "1h",  // window
                "namespace:\"kube-system\"+controllerKind:\"ReplicaSet\"+label[app]:\"ack-cost-exporter\"",  // filter
                null,  // step
                null,  // aggregate
                null,  // idle
                null,  // shareIdle
                null,  // shareSplit
                null,  // idleByNode
                null   // format
            );
            
            System.out.println("Successfully retrieved cost data");
            System.out.println("Response data structure:");
            
            if (response != null && response.getData() != null) {
                System.out.println("   - Number of data entries: " + response.getData().size());
                
                // Display each time range's data
                for (int i = 0; i < response.getData().size(); i++) {
                    Object dataItem = response.getData().get(i);
                    System.out.println("\n   Time Range " + (i + 1) + ":");
                    
                    if (dataItem instanceof java.util.Map) {
                        @SuppressWarnings("unchecked")
                        java.util.Map<String, Object> itemMap = (java.util.Map<String, Object>) dataItem;
                        
                        // Display each pod's cost data in this time range
                        for (java.util.Map.Entry<String, Object> entry : itemMap.entrySet()) {
                            String resourceName = entry.getKey();
                            Object allocationObj = entry.getValue();
                            
                            System.out.println("\n     Pod: " + resourceName);
                            
                            if (allocationObj instanceof org.openapitools.client.model.Allocation) {
                                org.openapitools.client.model.Allocation allocation = (org.openapitools.client.model.Allocation) allocationObj;
                                
                                // Display basic allocation info
                                if (allocation.getName() != null) {
                                    System.out.println("       Name: " + allocation.getName());
                                }
                                if (allocation.getStart() != null) {
                                    System.out.println("       Start: " + allocation.getStart());
                                }
                                if (allocation.getEnd() != null) {
                                    System.out.println("       End: " + allocation.getEnd());
                                }
                                
                                // Display resource usage
                                if (allocation.getCpuCoreRequestAverage() != null) {
                                    System.out.println("       CPU Request Average: " + String.format("%.2f", allocation.getCpuCoreRequestAverage()) + " cores");
                                }
                                if (allocation.getCpuCoreUsageAverage() != null) {
                                    System.out.println("       CPU Usage Average: " + String.format("%.2f", allocation.getCpuCoreUsageAverage()) + " cores");
                                }
                                if (allocation.getRamByteRequestAverage() != null) {
                                    System.out.println("       RAM Request Average: " + String.format("%.2f", allocation.getRamByteRequestAverage() / (1024 * 1024)) + " MB");
                                }
                                if (allocation.getRamByteUsageAverage() != null) {
                                    System.out.println("       RAM Usage Average: " + String.format("%.2f", allocation.getRamByteUsageAverage() / (1024 * 1024)) + " MB");
                                }
                                
                                // Display cost information
                                if (allocation.getCost() != null) {
                                    System.out.println("       Cost: " + String.format("%.4f", allocation.getCost()));
                                }
                                if (allocation.getCostRatio() != null) {
                                    System.out.println("       Cost Ratio: " + String.format("%.2f", allocation.getCostRatio() * 100) + "%");
                                }
                                if (allocation.getCustomCost() != null) {
                                    System.out.println("       Custom Cost: " + String.format("%.4f", allocation.getCustomCost()));
                                }
                                
                                // Display properties if available
                                org.openapitools.client.model.AllocationProperties properties = allocation.getProperties();
                                if (properties != null) {
                                    System.out.println("       Properties:");
                                    if (properties.getNamespace() != null) {
                                        System.out.println("         Namespace: " + properties.getNamespace());
                                    }
                                    if (properties.getNode() != null) {
                                        System.out.println("         Node: " + properties.getNode());
                                    }
                                    if (properties.getController() != null) {
                                        System.out.println("         Controller: " + properties.getController());
                                    }
                                    if (properties.getControllerKind() != null) {
                                        System.out.println("         Controller Kind: " + properties.getControllerKind());
                                    }
                                    if (properties.getCluster() != null) {
                                        System.out.println("         Cluster: " + properties.getCluster());
                                    }
                                }
                            } else if (allocationObj instanceof java.util.Map) {
                                @SuppressWarnings("unchecked")
                                java.util.Map<String, Object> allocation = (java.util.Map<String, Object>) allocationObj;
                                
                                // Display basic allocation info
                                if (allocation.get("name") != null) {
                                    System.out.println("       Name: " + allocation.get("name"));
                                }
                                if (allocation.get("start") != null) {
                                    System.out.println("       Start: " + allocation.get("start"));
                                }
                                if (allocation.get("end") != null) {
                                    System.out.println("       End: " + allocation.get("end"));
                                }
                                
                                // Display resource usage
                                Object cpuRequestObj = allocation.get("cpuCoreRequestAverage");
                                if (cpuRequestObj != null) {
                                    System.out.println("       CPU Request Average: " + String.format("%.2f", ((Number) cpuRequestObj).doubleValue()) + " cores");
                                }
                                Object cpuUsageObj = allocation.get("cpuCoreUsageAverage");
                                if (cpuUsageObj != null) {
                                    System.out.println("       CPU Usage Average: " + String.format("%.2f", ((Number) cpuUsageObj).doubleValue()) + " cores");
                                }
                                Object ramRequestObj = allocation.get("ramByteRequestAverage");
                                if (ramRequestObj != null) {
                                    System.out.println("       RAM Request Average: " + String.format("%.2f", ((Number) ramRequestObj).doubleValue() / (1024 * 1024)) + " MB");
                                }
                                Object ramUsageObj = allocation.get("ramByteUsageAverage");
                                if (ramUsageObj != null) {
                                    System.out.println("       RAM Usage Average: " + String.format("%.2f", ((Number) ramUsageObj).doubleValue() / (1024 * 1024)) + " MB");
                                }
                                
                                // Display cost information
                                Object costObj = allocation.get("cost");
                                if (costObj != null) {
                                    System.out.println("       Cost: " + String.format("%.4f", ((Number) costObj).doubleValue()));
                                }
                                Object costRatioObj = allocation.get("costRatio");
                                if (costRatioObj != null) {
                                    System.out.println("       Cost Ratio: " + String.format("%.2f", ((Number) costRatioObj).doubleValue() * 100) + "%");
                                }
                                Object customCostObj = allocation.get("customCost");
                                if (customCostObj != null) {
                                    System.out.println("       Custom Cost: " + String.format("%.4f", ((Number) customCostObj).doubleValue()));
                                }
                                
                                // Display properties if available
                                @SuppressWarnings("unchecked")
                                java.util.Map<String, Object> properties = (java.util.Map<String, Object>) allocation.get("properties");
                                if (properties != null) {
                                    System.out.println("       Properties:");
                                    if (properties.get("namespace") != null) {
                                        System.out.println("         Namespace: " + properties.get("namespace"));
                                    }
                                    if (properties.get("node") != null) {
                                        System.out.println("         Node: " + properties.get("node"));
                                    }
                                    if (properties.get("controller") != null) {
                                        System.out.println("         Controller: " + properties.get("controller"));
                                    }
                                    if (properties.get("controllerKind") != null) {
                                        System.out.println("         Controller Kind: " + properties.get("controllerKind"));
                                    }
                                    if (properties.get("cluster") != null) {
                                        System.out.println("         Cluster: " + properties.get("cluster"));
                                    }
                                }
                            }
                        }
                    }
                }
            } else {
                System.out.println("   - No matching resources found");
            }
            
        } catch (IOException e) {
            System.err.println("Kubeconfig loading failed: " + e.getMessage());
            System.err.println("Please ensure:");
            System.err.println("  1. KUBECONFIG environment variable is set, or");
            System.err.println("  2. kubeconfig file exists at ~/.kube/config");
            System.exit(1);
        } catch (ApiException e) {
            System.err.println("API call failed: " + e.getMessage());
            System.err.println("Status code: " + e.getCode());
            System.err.println("Response body: " + e.getResponseBody());
            System.exit(1);
        } catch (Exception e) {
            System.err.println("Other error: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
        
        System.out.println("============================================================");
        System.out.println("Example completed successfully!");
    }
}